src/Compilers/Core/Portable/Compilation.EmitStreamProvider.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// The compilation object is an immutable representation of a single invocation of the
+    /// compiler. Although immutable, a compilation is also on-demand, and will realize and cache
+    /// data as necessary. A compilation can produce a new compilation from existing compilation
+    /// with the application of small deltas. In many cases, it is more efficient than creating a
+    /// new compilation from scratch, as the new compilation can reuse information from the old
+    /// compilation.
+    /// </summary>
+    public abstract partial class Compilation
+    {
+        /// <summary>
+        /// Abstraction that allows the caller to delay the creation of the <see cref="Stream"/> values 
+        /// until they are actually needed.
+        /// </summary>
+        internal abstract class EmitStreamProvider
```

Comment:
I think it should be fine to make this change in behavior. Strictly speaking the behavior won't change much because we are calling DocumentationCommentCompiler.WriteDocumentationCommentXml before all error conditions are checked. From my perspective, the goal of using StreamProvider for XML Doc file is not to prevent the file from being written in presence of errors, but to be able to close it immediately after the method is called.